File: tests/test_send_receive_final.cpp
//
// Created by Ivan Shynkarenka on 09.07.2018
//

#include "test.h"

#include "../proto/protoex_final_protocol.h"
#include "../proto/test_final_protocol.h"

class MyFinalSender : public FBE::protoex::FinalSender
{
protected:
    size_t onSend(const void* data, size_t size) override
    {
        // Send nothing...
        return 0;
    }
};

class MyFinalReceiver : public FBE::protoex::FinalReceiver
{
public:
    MyFinalReceiver() : _order(false), _balance(false), _account(false) {}

    bool check() const { return _order && _balance && _account; }

protected:
    void onReceive(const protoex::OrderMessage& value) override { _order = true; }
    void onReceive(const protoex::BalanceMessage& value) override { _balance = true; }
    void onReceive(const protoex::AccountMessage& value) override { _account = true; }

private:
    bool _order;
    bool _balance;
    bool _account;
};

bool SendAndReceiveFinal(size_t index1, size_t index2)
{
    MyFinalSender sender;

    // Create and send a new order
    protoex::OrderMessage order({ protoex::Order({ 1, "EURUSD", protoex::OrderSide::buy, protoex::OrderType::market, 1.23456, 1000.0, 0.0, 0.0 }) });
    sender.send(order);

    // Create and send a new balance wallet
    protoex::BalanceMessage balance({ protoex::Balance({ "USD", 1000.0 }, 100.0) });
    sender.send(balance);

    // Create and send a new account with some orders
    protoex::AccountMessage account({ protoex::Account({ 1, "Test", protoex::StateEx::good, { proto::Balance("USD", 1000.0), 100.0 }, std::make_optional<protoex::Balance>({ proto::Balance("EUR", 100.0), 10.0 }), {} }) });
    account.body.orders.emplace_back(1, "EURUSD", protoex::OrderSide::buy, protoex::OrderType::market, 1.23456, 1000.0, 0.0, 0.0);
    account.body.orders.emplace_back(2, "EURUSD", protoex::OrderSide::sell, protoex::OrderType::limit, 1.0, 100.0, 0.0, 0.0);
    account.body.orders.emplace_back(3, "EURUSD", protoex::OrderSide::buy, protoex::OrderType::stop, 1.5, 10.0, 0.0, 0.0);
    sender.send(account);

    MyFinalReceiver receiver;

    // Receive data from the sender
    index1 %= sender.buffer().size();
    index2 %= sender.buffer().size();
    index2 = std::max(index1, index2);
    receiver.receive(sender.buffer().data(), index1);
    receiver.receive(sender.buffer().data() + index1, index2 - index1);
    receiver.receive(sender.buffer().data() + index2, sender.buffer().size() - index2);
    return receiver.check();
}

TEST_CASE("Send & Receive (Final)", "[FBE]")
{
    for (size_t i = 0; i < 1000; ++i)
        for (size_t j = i; j < 1000; ++j)
            REQUIRE(SendAndReceiveFinal(i, j));
}

File: projects/JavaScript/tests/test_send_receive_final.js
/* eslint-disable prefer-const,no-loss-of-precision */
'use strict'

const test = require('tape')

const proto = require('../proto/proto')
const protoex = require('../proto/protoex')

class MyFinalSender extends protoex.FinalSender {
  // noinspection JSUnusedLocalSymbols
  onSend (buffer, offset, size) {
    //  Send nothing...
    return 0
  }
}

class MyFinalReceiver extends protoex.FinalReceiver {
  constructor () {
    super()
    this._order = false
    this._balance = false
    this._account = false
  }

  check () {
    return (this._order && this._balance && this._account)
  }

  // noinspection JSUnusedLocalSymbols
  onReceive_OrderMessage (value) { this._order = true } // eslint-disable-line
  // noinspection JSUnusedLocalSymbols
  onReceive_BalanceMessage (value) { this._balance = true } // eslint-disable-line
  // noinspection JSUnusedLocalSymbols
  onReceive_AccountMessage (value) { this._account = true } // eslint-disable-line
}

function sendAndReceiveFinal (index1, index2) {
  let sender = new MyFinalSender()

  // Create and send a new order
  let order = new protoex.Order(1, 'EURUSD', protoex.OrderSide.buy, protoex.OrderType.market, 1.23456, 1000.0, 100.0)
  sender.send(new protoex.OrderMessage(order))

  // Create and send a new balance wallet
  let balance = new protoex.Balance(new proto.Balance('USD', 1000.0), 100.0)
  sender.send(new protoex.BalanceMessage(balance))

  // Create and send a new account with some orders
  let account = new protoex.Account(1, 'Test', protoex.StateEx.good, new protoex.Balance(new proto.Balance('USD', 1000.0), 100.0), new protoex.Balance(new proto.Balance('EUR', 100.0), 10.0))
  account.orders.push(new protoex.Order(1, 'EURUSD', protoex.OrderSide.buy, protoex.OrderType.market, 1.23456, 1000.0, 0.0, 0.0))
  account.orders.push(new protoex.Order(2, 'EURUSD', protoex.OrderSide.sell, protoex.OrderType.limit, 1.0, 100.0, 0.0, 0.0))
  account.orders.push(new protoex.Order(3, 'EURUSD', protoex.OrderSide.buy, protoex.OrderType.stop, 1.5, 10.0, 0.0, 0.0))
  sender.send(new protoex.AccountMessage(account))

  let receiver = new MyFinalReceiver()

  // Receive data from the sender
  index1 %= sender.buffer.size
  index2 %= sender.buffer.size
  index2 = Math.max(index1, index2)
  receiver.receive(sender.buffer, 0, index1)
  receiver.receive(sender.buffer, index1, index2 - index1)
  receiver.receive(sender.buffer, index2, sender.buffer.size - index2)
  return receiver.check()
}

test('Send & Receive (Final)', function (t) {
  for (let i = 0; i < 100; i++) {
    for (let j = 0; j < 100; j++) {
      t.true(sendAndReceiveFinal(i, j))
    }
  }
  t.end()
})
